Digipost/APIClient.swift

```diff
@@ -169,25 +177,77 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
         }
     }
 
-    func logout () {
-        logout(success: { () -> Void in
-            OAuthToken.removeAllTokens()
-            POSModelManager.sharedManager().deleteAllObjects()
-            }) { (error) -> () in
-                OAuthToken.removeAllTokens()
-                POSModelManager.sharedManager().deleteAllObjects()
+    func logoutThenDeleteAllStoredData() {
+        cancelAllRunningTasks { () -> Void in
+            self.logout(success: { () -> Void in
+                // get run for every time a sucessful scope logs out
+                }) { (error) -> () in
+                // gets run for every failed request
+            }
         }
     }
 
+    /**
+    success and failure blocks are run multiple times depending on how many requests are done
+    logs out OAuth tokens for all scopes in storage
+
+    */
     private func logout(#success: () -> Void, failure: (error: APIError) -> ()) {
         let rootResource = POSRootResource.existingRootResourceInManagedObjectContext(POSModelManager.sharedManager().managedObjectContext)
         if rootResource == nil {
             success()
             return
         }
-        validateFullScope {
-            let task = self.urlSessionTask(httpMethod.post, url: rootResource.logoutUri, success: success, failure: failure)
+        let logoutURI = rootResource.logoutUri
+        // if validation fails, just delete everything to make sure user will get correctly logged out in app
+        validateFullScope(success: {
+            let task = self.urlSessionTask(httpMethod.post, url: logoutURI, success: success, failure: failure)
             task.resume()
+
+            if let fullHighAuthToken = OAuthToken.oAuthTokenWithScope(kOauth2ScopeFullHighAuth) {
```

Comment:
Just so I understand this correctly: if validate fails, we just proceed to invalidate the next token?